### DISCUSSION TITLE
타입힌트를 어떻게 사용하면 좋을까요?

### DISCUSSION THREAD

**Discussion by Kirade** [8percent/styleguide#47] [General]
이전 논의에서 타입 힌트에 대해서 어떻게 생각하는지 논의를 진행한적이 있습니다.
https://github.com/8percent/styleguide/discussions/18

최근에 많은 분들이 타입힌팅을 활용해주시기 시작했는데요.
기존에 코드들이 타입힌트가 있지 않다보니, 현재 저희 코드들에 듬성듬성 타입힌팅을 사용하고 있고 저마다 타입힌트를 활용하는 범위 같은것들이 달라서 코드 스타일의 일관성이 떨어져가고 있는 것으로 보입니다. 일관성이 없는 코드는 가독성을 떨어트려 저희의 생산성에 영향을 주는 부분이라고 생각합니다.

개인적으로는 타입 힌트는 클래스나 함수의 매개변수 반환값 등을 정도를 표기하는 방식으로 가볍게 사용하고 있는데요.
다른분들은 어떤 수준으로 사용하고 계신지도 궁금하고, 어떤 방식을 저희 코드의 규칙으로 만들면 좋을지 의견을 부탁드립니다.

**Comment by jseop-lim** [comment]
## 타입힌트 사용 목적
가독성 개선 및 IDE 기능 보완을 목적으로 우리 코드에서 타입 힌트를 사용 중입니다. 이는 다음 상황에 근거합니다.
* 정적 검사기가 작동하지 않고 있습니다. 타입 힌팅이 적용되지 않은 코드가 대다수이므로 정적 검사 적용 시점은 매우 뒤로 예상합니다.
* 언급한 가독성이란 구체적으로 객체의 정의부를 보지 않고도 객체를 사용 가능한 상태를 말합니다.
  * [타입힌트를 통해 가독성 문제를 해결한 예시 PR](https://github.com/8percent/eight/pull/15527)
* 적절한 타입힌트를 통해 IDE의 Go to Declaration 기능(Pycharm에서 command+B)이 작동하지 않는 코드를 해결할 수 있습니다.
  * 함수의 return type 지정은 여기에 해당하는 좋은 예시입니다.

## ⭐ 타입힌트 적용 기준 및 예시
**아래 두 수준 사이에서 타입힌트를 붙이고 있습니다. 모범 사례 [코드](https://github.com/8percent/eight/blob/5886cebf93548f00091b9614bf4a42b739eb4e24/apps/communications/notifications/operations.py#L210) 링크를 첨부합니다.**

* 최소 기준: IDE의 Go to Declaration 기능(Pycharm에서 command+B)이 적용되는 수준
* 최대 기준: 타입힌트 명시가 가능한 모든 장소

주요 사용 시점은 다음과 같습니다. (생각나면 더 추가하겠습니다.)

### 함수의 매개변수와 반환값에는 반드시 타입힌트 사용

### 나중에 원소를 채우기 위해 미리 정의하는 딕셔너리나 리스트
> 아래 예시에서 errors 지역변수를 확인해주세요.
```python
def validate_advertisement_body(cls, body: str):
    """광고성 앱 푸시의 본문을 검증합니다."""
    errors: list[InvalidAppPushText] = []
    if not body.startswith('(광고)'):
        errors.append(InvalidAppPushText('광고성 앱 푸시의 본문은 "(광고)"로 시작해야 합니다.'))
    if not body.endswith(INDICATION_OF_REJECTING_AD_APP_PUSH):
        errors.append(InvalidAppPushText(f'광고성 앱 푸시의 본문은 "{INDICATION_OF_REJECTING_AD_APP_PUSH}"으로 끝나야 합니다.'))
    if errors:
        raise InvalidAppPushText(errors)
```

### 다른 함수를 호출하여 반환된 값을 저장하는 지역 변수에는 타입힌트 사용
```python
daily_payment_stat: DailyPayment = stat_maker.make_daily_payment_stat()
```

### IDE에서 타입을 유추하지 못하는 경우

#### Django에서 related_name을 통한 OneToOneField 역참조
```python
my_schedule: Schedule = deal.expected_schedule
```

#### 테스트 코드에서 Factory 객체 생성
```python
test_deal: Deal = DealFactory()
```

**Reply by jseop-lim** [reply]
재할당은 제외한다. (한 번 선언된 타입은 유지되어야 한다는 가정 아래에서)

**Comment by jseop-lim** [comment]
## 의견 나누기

### 정적 타입 검사기 도입 불가 이유
* Pycharm 플러그인 부재
* Django 자체가 타입이 없으므로 Django 추론이 어려움 - [Django Stub](https://github.com/typeddjango/django-stubs)으로 보완 가능한지 검증 필요
* Django Stub를 온전히 적용하기 위해선 Django 4.0 이상 업데이트 필요

타입을 쓰지 않은 코드를 소급하여 검사하진 않게 설정 가능하나, 자동 추론 오류는 해결이 필요합니다.

@cbscsm 
Django Stub 도입을 검토합니다. 실제 적용 가능할지 테스트가 필요합니다. 궁극적으로 정적 검사기를 도입하는 방향으로 갑니다. 사람이 타입을 모두 검사하기엔 한계가 있기 때문입니다. 타입힌트는 기계가 읽는 것입니다. 자동 타입추론을 돕는 범위에서 사용하고, 사람이 읽기 위한 타입힌트는 과합니다.

@Kirade 
정적 검사기 없는 모호한 상태를 유지하기보다, 아예 검사기를 도입하거나 타입힌트 안 쓰는 것이 좋을 것 같습니다. 중간 상태에서는 올바르지 않은 타입힌트가 난립할 수 있습니다.

@donghyun0218 
코드가 길어지므로 변수 할당에까지 타입힌트를 다는 것은 가독성을 저해합니다. 함수 매개변수에 동의합니다.


## 의견 모으기
### 타입힌트의 목적?
* 정적 검사기 - 아직 보류
* IDE 도움, 가독성 - 사람마다 다름 (전제)

### 타입힌트의 범위?
* 함수 매개변수와 반환 형태
* 변수 초기화

## 정책
* 각자 알아서 틀리지 않게 쓰기 - 틀리지 않는 한 작성자의 취향 존중하여 리뷰
* 최소 기준 정하기 - 함수 매개변수와 반환 형태
* 최대 기준 정하기 - 없음

**Comment by jseop-lim** [comment]
#59 에서 가이드 반영

**Comment by jseop-lim** [comment]
### 테스트 코드 Fixture의 타입힌트

✅ **Fixture에는 타입힌트를 적지 않는다.**

* Fixture에 타입힌트를 적지 않아도 IDE는 자동완성 기능을 제공한다.
* pytest 개발진들도 pytest에 대한 별도의 타이핑 기능을 고려하진 않는다.

**Comment by jseop-lim** [comment]
### 스타일 가이드의 타입힌트 규정 논의

✅ **일단 스타일 가이드 규칙을 개정하고, 타입힌트 활용은 자율에 맡기자.**

* 일괄 적용하거나 다 안쓰는 게 좋을 것 같다. 일괄 적용할 때는 정적 검사기를 반드시 도입해야 한다.
* 정적 검사기 도입과 일괄 적용으로 인한 코드 리팩터링 작업량에 비해 얻는 효용이 크지 않아 보인다.
* Django에서의 타입힌트에 대한 Best Practice를 찾아보자.
* 현재 등록된 단순한 규칙(함수 매개변수와 반환형에 반드시 표기)으로는 모든 경우를 설명할 수 없다.
